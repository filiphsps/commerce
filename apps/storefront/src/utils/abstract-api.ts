/* eslint-disable unused-imports/no-unused-vars */
import type { Nullable, OnlineShop } from '@nordcom/commerce-db';

import { DocumentTransform } from '@apollo/client';
import { visit } from 'graphql';

import type { Locale } from '@/utils/locale';
import type { ApolloClient, TypedDocumentNode } from '@apollo/client';

export type OmitTypeName<T> = Omit<T, '__typename'>;

export type ApiOptions = { api: AbstractApi };

export type AbstractApi<Q = any> = {
    locale: () => Locale;
    shop: () => OnlineShop;
    query: <T>(
        query: Q,
        variables?: Record<string, string | number | boolean | object | Array<string | number | object> | null>,
        options?: {
            fetchPolicy?: RequestCache;
            tags?: string[];
            revalidate?: number;
        }
    ) => Promise<{ data: T | null; errors: readonly any[] | undefined }>;
};
export type AbstractApiBuilder<K, Q> = ({
    api,
    locale,
    shop
}: {
    api: K;
    locale: Locale;
    shop: OnlineShop;
}) => AbstractApi<Q>;

export type AbstractShopifyApolloApiBuilder<Q> = AbstractApiBuilder<ApolloClient<any>, Q>;

export function buildCacheTagArray(shop: OnlineShop, locale: Locale, tags: string[]) {
    // TODO: change `shopify` tag based on the api we're using.
    return ['shopify', `shopify.${shop.id}`, shop.domain, locale.code, ...tags];
}

/**
 * Creates an AbstractApiBuilder for Shopify Apollo APIs.
 *
 * @todo TODO: Improve the type safety of all `AbstractApi` implementations.
 *
 * @param {object} options - The api options.
 * @param {ApolloClient<any>} options.api - The Apollo client to use.
 * @param {Locale} options.locale - The locale to use.
 * @param {OnlineShop} options.shop - The locale to use.
 * @returns {AbstractApiBuilder} The AbstractApiBuilder.
 */
export const ApiBuilder: AbstractShopifyApolloApiBuilder<TypedDocumentNode<any, any>> = ({ api, locale, shop }) => ({
    locale: () => locale,
    shop: () => shop,
    query: async (query, variables = {}, { tags = [], revalidate = undefined, fetchPolicy = undefined } = {}) => {
        const { data, errors, error } = await api.query({
            query,
            //fetchPolicy,
            context: {
                fetchOptions: {
                    cache: fetchPolicy ?? 'no-cache',
                    next: {
                        revalidate: revalidate ?? undefined,
                        tags: buildCacheTagArray(shop, locale, tags)
                    }
                }
            },
            variables: {
                language: locale.language,
                country: locale.country,
                ...variables
            }
        });

        return { data: data || null, errors, error };
    }
});

/**
 * @todo TODO: This should be replaced with a generalized shopify parser.
 *       Preferably one that we can use to output in whatever format we want.
 */
export const cleanShopifyHtml = (html: string | unknown): Nullable<string> => {
    if (typeof html !== 'string' || !html) return null;
    let out = (html as string) || '';

    // Remove all non-breaking spaces and replace them with normal spaces.
    // TODO: This is a hacky solution. We should write a proper shopify parser.
    out = out.replaceAll(/ /g, ' ').replaceAll('\u00A0', ' ');

    // Replace some of the more common unicode characters with their HTML.
    out = out
        .replaceAll('”', '&rdquo;')
        .replaceAll('“', '&ldquo;')
        .replaceAll('‘', '&lsquo;')
        .replaceAll('’', '&rsquo;')
        .replaceAll('…', '&hellip;');

    // Trim the preceding and trailing whitespace.
    out = out.trim();

    return out;
};

/**
 * A GraphQL document transform that adds the `@inContext` directive to all queries.
 *
 * THIS IS (or at least seems to be) A REALLY HACKY SOLUTION. USE WITH CAUTION. :^)
 *
 * @todo TODO: This should be placed in a separate file or even a separate package.
 * @todo TODO: Need to add tests for this.
 * @todo TODO: This currently adds duplicate variables if they already exist.
 */
export const shopifyContextTransform = new DocumentTransform((document) => {
    // Add Shopify's `@inContext` directive to all queries.
    const transformedDocument = visit(document, {
        OperationDefinition(node) {
            return {
                ...node,
                variableDefinitions: [
                    ...(node.variableDefinitions || []),
                    {
                        kind: 'VariableDefinition',
                        variable: {
                            kind: 'Variable',
                            name: {
                                kind: 'Name',
                                value: 'country'
                            }
                        },
                        type: {
                            kind: 'NamedType',
                            name: {
                                kind: 'Name',
                                value: 'CountryCode'
                            }
                        }
                    },
                    {
                        kind: 'VariableDefinition',
                        variable: {
                            kind: 'Variable',
                            name: {
                                kind: 'Name',
                                value: 'language'
                            }
                        },
                        type: {
                            kind: 'NamedType',
                            name: {
                                kind: 'Name',
                                value: 'LanguageCode'
                            }
                        }
                    }
                ],
                directives: [
                    ...(node.directives || []),
                    {
                        kind: 'Directive',
                        name: {
                            kind: 'Name',
                            value: 'inContext'
                        },
                        arguments: [
                            {
                                kind: 'Argument',
                                name: {
                                    kind: 'Name',
                                    value: 'country'
                                },
                                value: {
                                    kind: 'Variable',
                                    name: {
                                        kind: 'Name',
                                        value: 'country'
                                    }
                                }
                            },
                            {
                                kind: 'Argument',
                                name: {
                                    kind: 'Name',
                                    value: 'language'
                                },
                                value: {
                                    kind: 'Variable',
                                    name: {
                                        kind: 'Name',
                                        value: 'language'
                                    }
                                }
                            }
                        ]
                    }
                ]
            };
        }
    });
    return transformedDocument;
});

export type ApiReturn<T> = [T, undefined] | [undefined, Error];
