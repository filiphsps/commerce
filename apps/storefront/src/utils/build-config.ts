/* c8 ignore start */
export const BuildConfig = {
    environment: process.env.NODE_ENV,

    shopify: {
        api: '2024-07' as const
    }
};
/* c8 ignore stop */
