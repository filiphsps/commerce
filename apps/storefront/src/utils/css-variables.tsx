import { Shop } from '@nordcom/commerce-db';
import { TodoError } from '@nordcom/commerce-errors';

import { colord, extend } from 'colord';
import a11yPlugin from 'colord/plugins/a11y';

extend([a11yPlugin]);

// TODO: Generalize this
export const getBrandingColors = async (domain: string) => {
    const shop = await Shop.findByDomain(domain, { sensitiveData: true });
    if (shop.design.accents.length <= 0) {
        throw new TodoError();
    }
    const accents = shop.design.accents;

    // TODO: Deal with variants.
    const primary = accents
        .filter(({ type }) => type === 'primary')
        .sort((a, b) => (colord(a.color).luminance() < colord(b.color).luminance() ? -1 : 1))[0];
    const secondary = accents
        .filter(({ type }) => type === 'secondary')
        .sort((a, b) => (colord(a.color).luminance() < colord(b.color).luminance() ? -1 : 1))[0];

    return {
        primary,
        secondary
    };
};

const CssVariablesProvider = async ({ domain }: { domain: string }) => {
    const branding = await getBrandingColors(domain);

    // TODO: Background and foreground colors.
    return (
        <style suppressHydrationWarning={true}>{`
        :root {
            --color-background: #fefefe;
            --color-foreground: #101418;

            --color-accent-primary: ${branding.primary.color};
            --color-accent-primary-text: ${branding.primary.foreground};
            --color-accent-primary-light: ${colord(branding.primary.color).lighten(0.115).saturate(0.15).toHex()};
            --color-accent-primary-dark: ${colord(branding.primary.color).darken(0.05).toHex()};

            --color-accent-secondary: ${branding.secondary.color};
            --color-accent-secondary-text: ${branding.secondary.foreground};
            --color-accent-secondary-light: ${colord(branding.secondary.color).lighten(0.195).saturate(0.15).toHex()};
            --color-accent-secondary-dark: ${colord(branding.secondary.color).darken(0.15).toHex()};

            /* TODO: Remove these legacy variables. */
            --accent-primary: var(--color-accent-primary);
            --accent-primary-light: var(--color-accent-primary-light);
            --accent-primary-dark: var(--color-accent-primary-dark);
            --accent-secondary: var(--color-accent-secondary);
            --accent-secondary-light: var(--color-accent-secondary-light);
            --accent-secondary-dark: var(--color-accent-secondary-dark);
        }
    `}</style>
    );
};

CssVariablesProvider.displayName = 'Nordcom.CssVariablesProvider';
export { CssVariablesProvider };
