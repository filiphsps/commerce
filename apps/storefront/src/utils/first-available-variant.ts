import { NotFoundError } from '@nordcom/commerce-errors';

import type { Product, ProductVariant } from '@/api/product';

/**
 * Find the first available and most suitable variant for a specific product.
 *
 * @param {Product} product - The product data.
 * @returns {ProductVariant} The recommended variant based on availability and price.
 */
export const firstAvailableVariant = (product?: Product | null): ProductVariant | undefined => {
    // 1. Make sure we got a product passed to us.
    if (!product?.variants) {
        return undefined;
    }

    const variants: ProductVariant[] =
        ((product.variants.edges as any)
            ? product.variants.edges.map(({ node: variant }) => variant)
            : (product.variants as any)) || [];

    if (variants.length <= 0) {
        throw new NotFoundError(`"product.variant"`);
    }

    // 2. Check if the last variant is available.
    if (variants.at(-1)?.availableForSale) {
        // 2.1. If it is, return it.
        return variants.at(-1)!;
    }
    // 2.2. If it isn't, continue.

    // 3. Loop through the variants in reverse to get the more expensive variants first.
    for (let i = variants.length - 1; i >= 0; i--) {
        const variant = variants[i] as (typeof variants)[number] | undefined;

        // 3.1. If the variant is not available, continue.
        if (!variant?.availableForSale) {
            continue;
        }

        // 3.2. If the variant is available, return it.
        return variant;
    }

    // 4. If no variants are available, return the most expensive one.
    return variants.at(-1)!;
};
