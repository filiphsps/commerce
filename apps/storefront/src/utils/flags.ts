/* c8 ignore start */
import { get } from '@vercel/edge-config';
import { unstable_flag as flag } from '@vercel/flags/next';

export const showSearchFilter = flag({
    key: 'search-filter',
    async decide() {
        const value = await get(this.key);
        return !!value || false;
    }
});

export const showProductInfoLines = flag({
    key: 'product-page-info-lines',
    async decide() {
        const value = await get(this.key);
        return !!value || false;
    }
});

export const showHeaderSearchBar = flag({
    key: 'header-search-bar',
    async decide() {
        const value = await get(this.key);
        return !!value || false;
    }
});
/* c8 ignore stop */
