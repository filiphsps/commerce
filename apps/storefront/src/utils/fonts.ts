import { Public_Sans } from 'next/font/google';

const primaryFont = Public_Sans({
    weight: 'variable',
    subsets: ['latin'],
    display: 'swap',
    variable: '--font-primary',
    preload: true
});

export { primaryFont };
