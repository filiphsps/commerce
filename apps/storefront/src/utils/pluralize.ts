/**
 * Pluralize a noun based on a count
 *
 * @param {object} options - The options.
 * @param {number} options.count - The count of the noun
 * @param {string} options.noun - The noun to pluralize
 * @param {string} [options.suffix='s'] - The suffix to add to the noun if the count is not 1
 * @returns {string} The pluralized noun
 */
export const pluralize = ({ count = 0, noun, suffix = 's' }: { count: number; noun: string; suffix?: string }) =>
    `${noun}${(count !== 1 && suffix) || ''}`;
