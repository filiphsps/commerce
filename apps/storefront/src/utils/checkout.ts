import type { OnlineShop } from '@nordcom/commerce-db';
import { InvalidCartError, UnknownCommerceProviderError } from '@nordcom/commerce-errors';

import { productToMerchantsCenterId } from '@/utils/merchants-center-id';
import { safeParseFloat } from '@/utils/pricing';

import type { Locale } from '@/utils/locale';
import type { TrackableContextValue } from '@/utils/trackable';
import type { CartWithActions } from '@shopify/hydrogen-react';
import type { CartLine } from '@shopify/hydrogen-react/storefront-api-types';

// Const hacky workaround for ga4 cross-domain
// taken from StackOverflow
export const getCrossDomainLinkerParameter = (domain: string) => {
    // create form element, give it an action, make it hidden and prevent the submit event
    const formNode = document.createElement('form') as any;
    formNode.action = `https://checkout.${domain}`; // TODO: This should be dependant on the tenant.
    formNode.style.opacity = '0';
    formNode.addEventListener('submit', (event: any) => {
        event.preventDefault();
    });

    // create a button node, make it type=submit and append it to the form
    const buttonNode = document.createElement('button') as any;
    buttonNode.type = 'submit';
    formNode.append(buttonNode);

    // append the form (and button) to the DOM
    document.body.append(formNode);

    // trigger a click on the button node to submit the form
    buttonNode.click();

    // check for the input[name=_gl] hidden input in the form (if decoration worked)
    const _glNode = formNode.querySelector('input[name="_gl"]') as any;

    if (_glNode) {
        return _glNode.value as string;
    }

    console.warn(`Could not find _gl input in checkout form with action "${formNode.action}"`);
    return null;
};

export const Checkout = async ({
    shop,
    locale,
    cart,
    trackable
}: {
    shop: OnlineShop;
    locale: Locale;
    cart: CartWithActions;
    trackable: TrackableContextValue;
}) => {
    if (shop.commerceProvider.type !== 'shopify') {
        throw new UnknownCommerceProviderError();
    }

    if (typeof (cart as any) === 'undefined' || !(cart as any)) {
        throw new InvalidCartError('Cart is undefined or null');
    } else if (!cart.totalQuantity || cart.totalQuantity <= 0 || !cart.lines) {
        throw new InvalidCartError('Cart is empty');
    } else if (!cart.checkoutUrl) {
        throw new InvalidCartError('Cart is missing checkoutUrl');
    }

    const url = cart.checkoutUrl.replace(/[A-Za-z0-9\-\_]+.\.myshopify\.com/, shop.commerceProvider.domain);

    try {
        trackable.postEvent('begin_checkout', {
            gtm: {
                ecommerce: {
                    currency: cart.cost?.totalAmount?.currencyCode!,
                    value: safeParseFloat(undefined, cart.cost?.totalAmount?.amount),
                    items: (cart.lines.filter((_) => _) as CartLine[]).map((line) => ({
                        item_id: productToMerchantsCenterId({
                            locale: locale,
                            product: {
                                productGid: line.merchandise!.product!.id,
                                variantGid: line.merchandise!.id
                            } as any
                        }),
                        item_name: line.merchandise.product.title,
                        item_variant: line.merchandise.title,
                        item_brand: line.merchandise.product.vendor,
                        item_category: line.merchandise.product.productType || undefined,
                        sku: line.merchandise.sku || undefined,
                        product_id: line.merchandise!.product!.id,
                        variant_id: line.merchandise!.id,
                        currency: line.merchandise.price.currencyCode!,
                        price: safeParseFloat(undefined, line.merchandise.price.amount),
                        quantity: line.quantity
                    }))
                }
            }
        });
    } catch (error: unknown) {
        console.error(error);
    }

    const ga4 = getCrossDomainLinkerParameter(shop.domain);
    window.location.href = `${url}${ga4 ? `${url.includes('?') ? '&' : '?'}_gl=${ga4}` : ''}`;
};
