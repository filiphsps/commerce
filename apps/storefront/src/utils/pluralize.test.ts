import { describe, expect, it } from 'vitest';

import { Pluralize } from '@/utils/pluralize';

describe('utils', () => {
    describe('Pluralize', () => {
        it(`should pluralize the noun when count is greater than 1`, () => {
            const count = 2;
            const noun = 'apple';
            const suffix = 's';

            const result = Pluralize({ count, noun, suffix });

            expect(result).toBe('apples');
            //expect(result).toMatchSnapshot();
        });

        it(`should not pluralize the noun when count is 1`, () => {
            const count = 1;
            const noun = 'apple';
            const suffix = 's';

            const result = Pluralize({ count, noun, suffix });

            expect(result).toBe('apple');
            //expect(result).toMatchSnapshot();
        });

        it(`should use the provided suffix when count is greater than 1`, () => {
            const count = 2;
            const noun = 'child';
            const suffix = 'ren';

            const result = Pluralize({ count, noun, suffix });

            expect(result).toBe('children');
            //expect(result).toMatchSnapshot();
        });

        it(`should not use the provided suffix when count is 1`, () => {
            const count = 1;
            const noun = 'child';
            const suffix = 'ren';

            const result = Pluralize({ count, noun, suffix });

            expect(result).toBe('child');
            //expect(result).toMatchSnapshot();
        });
    });
});
