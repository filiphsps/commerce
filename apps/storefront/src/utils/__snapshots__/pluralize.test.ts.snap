// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > Pluralize > should not pluralize the noun when count is 1 1`] = `"apple"`;

exports[`utils > Pluralize > should not use the provided suffix when count is 1 1`] = `"child"`;

exports[`utils > Pluralize > should pluralize the noun when count is greater than 1 1`] = `"apples"`;

exports[`utils > Pluralize > should use the provided suffix when count is greater than 1 1`] = `"children"`;
